data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt

```diff
@@ -6,54 +6,77 @@ import androidx.paging.PagingData
 import io.filmtime.data.api.tmdb.TmdbMoviesRemoteSource
 import io.filmtime.data.api.trakt.TraktSearchRemoteSource
 import io.filmtime.data.api.trakt.TraktSyncRemoteSource
+import io.filmtime.data.database.MovieDetailDao
 import io.filmtime.data.model.CreditItem
 import io.filmtime.data.model.GeneralError
 import io.filmtime.data.model.Result
 import io.filmtime.data.model.VideoDetail
 import io.filmtime.data.model.VideoThumbnail
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 internal class TmdbMovieRepositoryImpl @Inject constructor(
   private val tmdbMoviesRemoteSource: TmdbMoviesRemoteSource,
   private val traktMovieSearchRemoteSource: TraktSearchRemoteSource,
   private val traktSyncRemoteSource: TraktSyncRemoteSource,
+  private val movieDao: MovieDetailDao,
 ) : TmdbMovieRepository {
 
-  override suspend fun getMovieDetails(movieId: Int): Result<VideoDetail, GeneralError> {
-    return when (val result = tmdbMoviesRemoteSource.movieDetails(movieId)) {
-      is Result.Failure -> result
-      is Result.Success -> {
-        return when (val traktIdResult = traktMovieSearchRemoteSource.getByTmdbId(result.data.ids.tmdbId.toString())) {
-          is Result.Failure -> traktIdResult
-          is Result.Success -> {
-            val traktId = traktIdResult.data.toInt()
-            return when (val watched = traktSyncRemoteSource.getHistoryById(traktId.toString())) {
-              is Result.Failure -> result.run {
-                copy(
-                  data = data.copy(
-                    ids = data.ids.copy(
-                      traktId = traktId,
+  override suspend fun getMovieDetails(movieId: Int): Flow<Result<VideoDetail, GeneralError>> {
```

Comment:
@moallemi Could you please review this function and help me how to improve this function? 
It works right now but I want to only depend on database but first time we get movie it returns ```null``` and we show user an error, not loading.
And another thing is that ```GlobalScope.lunch```. I need to lunch a coroutine but not suspend my current function what is your suggestion?

Reply:
I think we do not need a GlobalScope for now. one implementation could be:

```kotlin
override suspend fun getMovieDetails(movieId: Int): Flow<Result<VideoDetail, GeneralError>> = flow {
    val movie = movieDao.getMovieByTmdbId(movieId).firstOrNull()
    if (movie != null) {
      emit(Result.Success(movie.toMovie()))
      // TODO: Check if movie is outdated and then fetch from network
      fetchMovieDetailsFromNetwork(movieId)
    } else {
      emit(fetchMovieDetailsFromNetwork(movieId))
    }
  }

  private suspend fun fetchMovieDetailsFromNetwork(movieId: Int): Result<VideoDetail, GeneralError> =
    when (val result = tmdbMoviesRemoteSource.movieDetails(movieId)) {
      is Result.Failure -> result
      is Result.Success -> {
        movieDao.storeMovie(result.data.toEntity())
        when (val traktIdResult = traktMovieSearchRemoteSource.getByTmdbId(result.data.ids.tmdbId.toString())) {
          is Result.Failure -> traktIdResult
          is Result.Success -> {
            val traktId = traktIdResult.data.toInt()
            when (val watched = traktSyncRemoteSource.getHistoryById(traktId.toString())) {
              is Result.Failure -> result.run {
                copy(
                  data = data.copy(
                    ids = data.ids.copy(
                      traktId = traktId,
                    ),
                  ),
                )
              }

              is Result.Success -> {
                result.run {
                  copy(
                    data = data.copy(
                      ids = data.ids.copy(
                        traktId = traktId,
                      ),
                      isWatched = watched.data,
                    ),
                  )
                }
              }
            }
          }
        }
      }
    }
```
We currently don't need to verify if the movie is outdated. We can postpone implementing this feature and incorporate a cache expiration of either one or five days at a later stage.